crates/red_knot_python_semantic/resources/mdtest/terminal_statements.md

```diff
@@ -0,0 +1,134 @@
+# Terminal statements
+
+## Introduction
+
+Terminal statements complicate a naive control-flow analysis.
+
+As a simple example:
+
+```py
+def f(cond: bool) -> str:
+    if cond:
+        x = "test"
+    else:
+        raise ValueError
+    return x
+
+def g(cond: bool):
+    if cond:
+        x = "test"
+        reveal_type(x)  # revealed: Literal["test"]
+    else:
+        x = "unreachable"
+        reveal_type(x)  # revealed: Literal["unreachable"]
+        raise ValueError
+    reveal_type(x)  # revealed: Literal["test"]
+```
+
+In `f`, we should be able to determine that the `else` branch ends in a terminal statement, and that
+the `return` statement can only be executed when the condition is true. We should therefore consider
+the reference always bound, even though `x` is only bound in the true branch.
+
+Similarly, in `g`, we should see that the assignment of the value `"unreachable"` can never be seen
+by the final `reveal_type`.
+
+## `return` is terminal
+
+```py
+def f(cond: bool) -> str:
+    if cond:
+        x = "test"
+    else:
+        return "early"
+    return x
+
+def g(cond: bool):
+    if cond:
+        x = "test"
+        reveal_type(x)  # revealed: Literal["test"]
+    else:
+        x = "unreachable"
+        reveal_type(x)  # revealed: Literal["unreachable"]
+        return
+    reveal_type(x)  # revealed: Literal["test"]
+```
+
+## `continue` is terminal within its loop scope
+
+```py
+def f(cond: bool) -> str:
+    while True:
+        if cond:
+            x = "test"
+        else:
+            continue
+        return x
+
+def g(cond: bool, i: int):
+    x = "before"
+    while i < 5:
+        if cond:
+            x = "loop"
+            reveal_type(x)  # revealed: Literal["loop"]
+        else:
+            x = "continue"
+            reveal_type(x)  # revealed: Literal["continue"]
+            continue
+        reveal_type(x)  # revealed: Literal["loop"]
+    reveal_type(x)  # revealed: Literal["before", "loop"]
+```
+
+## `break` is terminal within its loop scope
+
+```py
+def f(cond: bool) -> str:
+    while True:
+        if cond:
+            x = "test"
+        else:
+            break
+        return x
+    return "late"
```

Comment:
We could `return x` here as well and expect the `possibly-unresolved-reference` error?